### PR TITLE
Only override session.TargetString if it has not been set

### LINKED ISSUE #5132: USE statement not respected with JDBC MySQL driver and dbname in connection params
If you connect to mysql using the JDBC MySQL driver with a dbname in connection params then USE statements take no effect.

This doesn't affect the mysql CLI because it sends the `COM_INIT_DB` packet. The JDBC driver doesn't send the `COM_INIT_DB`, it uses the USE statement. As seen here:
https://github.com/mysql/mysql-connector-j/blob/36534fa273b4d7824a8668ca685465cf8eaeadd9/src/main/user-impl/java/com/mysql/cj/jdbc/ConnectionImpl.java#L2139

The issue is this:
https://github.com/vitessio/vitess/blob/60b4643725017fa809e11c0d00a81e5cc5498586/go/vt/vtgate/plugin_mysql_server.go#L149-L151

You can see that `c.SchemaName` overwrites `session.TargetString` for every call and that the mysql protocol ignores any `USE` statements.

I think the fix is to assign the value of `session.TargetString` to `c.SchemaName` but I'm not sure what the wider repercussions are...

### PR DESCRIPTION
**Pull request by tirsen**
This fixes #5132

Signed-off-by: Jon Tirsen <jontirsen@squareup.com>

### CODE REVIEW

File: go/vt/vtgate/mysql_protocol_test.go [hunk 1]
```diff
@@ -92,6 +93,35 @@ func TestMySQLProtocolStreamExecute(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer c.Close()
+
+	qr, err := c.ExecuteFetch("show vitess_target", 1, false)
+	if err != nil {
+		t.Error(err)
+	}
+	assert.Equal(t, "VARCHAR(\"@master\")", qr.Rows[0][0].String())
```
Comment (demmer) [inline go/vt/vtgate/mysql_protocol_test.go]: For completeness I think we should have tests that show the expected behavior and not just the results of `show vitess_target`. Specifically, when in `@master` mode, run a simple `select * from table`  query against tables that are in more than one keyspace to show that the multi keyspace routing works.

Then when we force it to be using `TestUnsharded` show that the above no longer works for the tables in other keyspaces. 

We should also prove the behavior is correct when (a) the initial ConnParams contains no DbName and (b) that when ConnParams contains `TestUnsharded` you can still override to another keyspace or to `@master` to get the multi-keyspace mode.

(a) a test case that things work as expected when we connect without an intial

File: go/vt/vtgate/plugin_mysql_server.go [hunk 2]
```diff
@@ -155,6 +155,10 @@ func (vh *vtgateHandler) ComQuery(c *mysql.Conn, query string, callback func(*sq
 	}
 	session, result, err := vh.vtg.Execute(ctx, session, query, make(map[string]*querypb.BindVariable))
 	c.ClientData = session
+	if session != nil {
+		// Make sure the outcomeo of any potential USE statements are saved into the connection state
+		c.SchemaName = session.TargetString
```
Comment (demmer) [inline go/vt/vtgate/plugin_mysql_server.go]: This shouldn't be needed any more (and to me seems incorrect).

File: go/vt/vtgate/plugin_mysql_server.go [hunk 3]
```diff
@@ -268,14 +272,18 @@ func (vh *vtgateHandler) ComStmtExecute(c *mysql.Conn, prepare *mysql.PrepareDat
...
 		session.TargetString = c.SchemaName
 	}
 	if session.Options.Workload == querypb.ExecuteOptions_OLAP {
 		err := vh.vtg.StreamExecute(ctx, session, prepare.PrepareStmt, prepare.BindVars, callback)
 		return mysql.NewSQLErrorFromError(err)
 	}
 	_, qr, err := vh.vtg.Execute(ctx, session, prepare.PrepareStmt, prepare.BindVars)
+	if session != nil {
+		// Make sure the outcomeo of any potential USE statements are saved into the connection state
+		c.SchemaName = session.TargetString
```
Comment (demmer) [inline go/vt/vtgate/plugin_mysql_server.go]: ditto

File: go/vt/vtgate/plugin_mysql_server.go (hunk 2 above)
Comment (tirsen) [reply go/vt/vtgate/plugin_mysql_server.go]: Oops yeah I thought I had reverted all of these. Will revert now.

Comment (tirsen) [reply go/vt/vtgate/plugin_mysql_server.go]: Ah I had just forgotten to push :-/

File: go/vt/vtgate/mysql_protocol_test.go (hunk 1 above)
Comment (tirsen) [reply go/vt/vtgate/mysql_protocol_test.go]: I couldn't figure out how to do what you asked for without reams of test set up code but I think I've added sufficient tests to prove that we're using the vitess target for routing. I query on the `@replica` target with only `master` tablets and it fails with the expected error message.

**Comment by tirsen** [comment]
Yeah I agree. I spent a few hours digging into the sandbox and it doesn't look straightforward and certainly not worth the effort for this particular change. It would be nice with a more useful sandbox though!

**Pull request merged**